### PR TITLE
Convert util package to Kotlin

### PR DESCRIPTION
**Pull request by ZacSweers**
Ref #1252

### CODE REVIEW

File: moshi-adapters/src/main/java/com/squareup/moshi/adapters/EnumJsonAdapter.java [hunk 1]
```diff
@@ -68,7 +68,7 @@
       nameStrings = new String[constants.length];
       for (int i = 0; i < constants.length; i++) {
         String constantName = constants[i].name();
-        nameStrings[i] = jsonName(constantName, enumType.getField(constantName));
+        nameStrings[i] = jsonName(enumType.getField(constantName), constantName);
```
Comment (ZacSweers) [inline moshi-adapters/src/main/java/com/squareup/moshi/adapters/EnumJsonAdapter.java:71]: I think this is a problem for code gen actually :(

Comment (ZacSweers) [reply moshi-adapters/src/main/java/com/squareup/moshi/adapters/EnumJsonAdapter.java:71]: actually should be ok

File: moshi/src/main/java/com/squareup/moshi/internal/Util.kt [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+        val original = toResolve
+        val ownerType: Type? = original.ownerType
+        val newOwnerType = ownerType?.let {
+          ownerType.resolve(context, contextRawType, visitedTypeVariables)
+        }
+        var changed = newOwnerType !== ownerType
+        var args = original.actualTypeArguments
+        var t = 0
+        val length = args.size
+        while (t < length) {
```
Comment (JvmName) [inline moshi/src/main/java/com/squareup/moshi/internal/Util.kt:249]: isn't this equivalent, or am I missing something
```kotlin
for(t in 0..args.size)
```

File: moshi/src/main/java/com/squareup/moshi/internal/Util.kt [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+  if (toResolve == rawType) {
+    return context
+  }
+
+  // we skip searching through interfaces if unknown is an interface
+  if (toResolve.isInterface) {
+    val interfaces = rawType.interfaces
+    var i = 0
+    val length = interfaces.size
+    while (i < length) {
```
Comment (JvmName) [inline moshi/src/main/java/com/squareup/moshi/internal/Util.kt:311]:
```suggestion
	for(i in 0..interfaces.size) {
```

File: moshi/src/main/java/com/squareup/moshi/internal/Util.kt [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+    throw RuntimeException("Failed to find the generated JsonAdapter class for $type", e)
+  } catch (e: NoSuchMethodException) {
+    if (possiblyFoundAdapter != null && type !is ParameterizedType && possiblyFoundAdapter.typeParameters.isNotEmpty()) {
+      throw RuntimeException(
+        "Failed to find the generated JsonAdapter constructor for '" +
+          type +
+          "'. Suspiciously, the type was not parameterized but the target class '" +
+          possiblyFoundAdapter.canonicalName +
+          "' is generic. Consider using " +
+          "Types#newParameterizedType() to define these missing type variables.",
```
Comment (JvmName) [inline moshi/src/main/java/com/squareup/moshi/internal/Util.kt:416]:
```suggestion
        "Failed to find the generated JsonAdapter constructor for '$type'. Suspiciously, the type was not parameterized but the target class '${possiblyFoundAdapter.canonicalName}' is generic. Consider using Types#newParameterizedType() to define these missing type variables."
```

**Review by JvmName — APPROVED** [review]

File: moshi/src/main/java/com/squareup/moshi/internal/Util.kt (hunk 2 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/internal/Util.kt:249]: Close! `for(t in args.indices) {`

File: moshi/src/main/java/com/squareup/moshi/internal/Util.kt (hunk 4 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/internal/Util.kt]: Did this locally

File: moshi/src/main/java/com/squareup/moshi/internal/Util.kt (hunk 3 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/internal/Util.kt]: Same thing with indices

**Pull request merged**